gitlabform/processors/project/project_job_token_scope_processor.py

```diff
@@ -0,0 +1,186 @@
+
+
+class ProjectJobTokenScopeProcessor(AbstractProcessor):
+    def __init__(self, gitlab: GitLab):
+        super().__init__("job_token_scope", gitlab)
+
+    def _process_configuration(self, project_and_group: str, configuration: dict):
+        job_token_config = configuration.get("job_token_scope", {})
+        verbose(f"Job Token Scope config: {job_token_config}")
+
+        project = self.gl.projects.get(project_and_group)
+        job_token_scope = project.job_token_scope.get()
+
+        self._process_limit_access_to_this_project_setting(
+            job_token_config, job_token_scope
+        )
+
+        job_token_scope.refresh()
+
+        allowlist_config = job_token_config.get("allowlist", {})
+        verbose(f"configuration allowlist: {allowlist_config}")
+
+        if allowlist_config:
+            verbose("Processing Job Token allowlist")
+
+            self._process_groups(job_token_scope, allowlist_config.get("groups", []))
+            job_token_scope.refresh()
+            self._process_projects(
+                project, job_token_scope, allowlist_config.get("projects", [])
+            )
+        else:
+            warning(
+                "Removing existing groups and projects from Allowlist, as none set in configuration"
+            )
+
+            self._remove_groups_from_allowlist(
+                job_token_scope, job_token_scope.groups_allowlist.list(), []
+            )
+            job_token_scope.refresh()
+            self._remove_projects_from_allowlist(
+                project, job_token_scope, job_token_scope.allowlist.list(), []
+            )
+
+        job_token_scope.refresh()
+
+    @staticmethod
+    def _process_limit_access_to_this_project_setting(
+        configuration: dict, job_token_scope: ProjectJobTokenScope
+    ):
+        limit_access_to_this_project: bool = configuration.get(
+            "limit_access_to_this_project", True
+        )
+        verbose(f"limit_access_to_this_project: {limit_access_to_this_project}")
+        job_token_scope.enabled = limit_access_to_this_project
+        job_token_scope.save()
+
+    def _process_projects(
+        self,
+        project: Project,
+        job_token_scope: ProjectJobTokenScope,
+        projects_allowlist: List,
+    ):
+        if not projects_allowlist:
+            warning(
+                "Process will remove existing projects from allowlist, as none set in configuration"
+            )
+
+        existing_allowlist = job_token_scope.allowlist.list()
+
+        target_project_ids = self._get_target_project_ids_from_config(
+            projects_allowlist
+        )
+
+        project_ids_to_add = self._get_ids_to_add_to_allowlist(
+            existing_allowlist, target_project_ids
+        )
+        for project_id in project_ids_to_add:
+            if project_id != project.id:
+                job_token_scope.allowlist.create({"target_project_id": project_id})
+                job_token_scope.save()
```

Comment:
Does it make sense to perform this after completing the `for` loop?

Reply:
For speed - yes, _but_ defensively programming, if for some reason the add to allowlist fails or the job running the gitlabform falls over after, I think it it is better to have saved a good state as we go?

Reply:
That was certainly my reasoning behind putting the save in the loop

Reply:
That's a good point. In that case, I suggest we add some debug logs and specify which project has been added to the allowlist.